File: dart/editor/tools/plugins/com.google.dart.command.analyze/src/com/google/dart/command/analyze/AnalyzerImpl.java
/*
 * Copyright (c) 2012, the Dart project authors.
 * 
 * Licensed under the Eclipse Public License v1.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.dart.command.analyze;

import com.google.dart.engine.AnalysisEngine;
import com.google.dart.engine.context.AnalysisContext;
import com.google.dart.engine.context.AnalysisException;
import com.google.dart.engine.element.CompilationUnitElement;
import com.google.dart.engine.element.LibraryElement;
import com.google.dart.engine.error.AnalysisError;
import com.google.dart.engine.error.ErrorSeverity;
import com.google.dart.engine.error.ErrorType;
import com.google.dart.engine.internal.context.AnalysisOptionsImpl;
import com.google.dart.engine.sdk.DartSdk;
import com.google.dart.engine.sdk.DirectoryBasedDartSdk;
import com.google.dart.engine.source.DartUriResolver;
import com.google.dart.engine.source.ExplicitPackageUriResolver;
import com.google.dart.engine.source.FileBasedSource;
import com.google.dart.engine.source.FileUriResolver;
import com.google.dart.engine.source.PackageUriResolver;
import com.google.dart.engine.source.Source;
import com.google.dart.engine.source.SourceFactory;
import com.google.dart.engine.source.SourceKind;
import com.google.dart.engine.source.UriKind;
import com.google.dart.engine.utilities.source.LineInfo;

import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Scans, parses, and analyzes a library.
 */
public class AnalyzerImpl {
  /**
   * The maximum number of sources for which AST structures should be kept in the cache.
   */
  private static final int MAX_CACHE_SIZE = 256;

  private static final HashMap<File, DirectoryBasedDartSdk> sdkMap = new HashMap<File, DirectoryBasedDartSdk>();

  /**
   * @return the new or cached instance of the {@link DartSdk} with the given directory.
   */
  private static DirectoryBasedDartSdk getSdk(File sdkDirectory, boolean useDart2jsPaths) {
    DirectoryBasedDartSdk sdk = sdkMap.get(sdkDirectory);
    if (sdk == null) {
      sdk = new DirectoryBasedDartSdk(sdkDirectory, useDart2jsPaths);
      sdkMap.put(sdkDirectory, sdk);
    }
    return sdk;
  }

  private AnalyzerOptions options;

  private DirectoryBasedDartSdk sdk;

  public AnalyzerImpl(AnalyzerOptions options) {
    this.options = options;
    this.sdk = getSdk(options.getDartSdkPath(), options.getUseDart2jsPaths());
  }

  /**
   * Treats the {@code sourceFile} as the top level library and analyzes the unit for warnings and
   * errors. The errors are added to the given list, and line information for all sources that have
   * errors is added to the given map.
   * 
   * @param sourceFile file to analyze
   * @param errors the list to which errors will be added
   * @param lineInfoMap the map to which line information will be added
   * @return the severity of the most severe error or warning
   */
  public ErrorSeverity analyze(File sourceFile, List<AnalysisError> errors,
      Map<Source, LineInfo> lineInfoMap) throws IOException, AnalysisException {
    if (sourceFile == null) {
      throw new IllegalArgumentException("sourceFile cannot be null");
    }

    // create options for context
    AnalysisOptionsImpl contextOptions = new AnalysisOptionsImpl();
    contextOptions.setCacheSize(MAX_CACHE_SIZE);
    contextOptions.setEnableAsync(options.getEnableAsync());
    contextOptions.setEnableEnum(options.getEnableEnum());
    contextOptions.setHint(!options.getDisableHints());

    // prepare AnalysisContext
    AnalysisContext context = AnalysisEngine.getInstance().createAnalysisContext();
    context.setSourceFactory(createSourceFactory(sourceFile));
    context.setAnalysisOptions(contextOptions);

    // prepare Source
    sourceFile = sourceFile.getAbsoluteFile();
    Source librarySource = new FileBasedSource(getUri(sourceFile), sourceFile);

    return performAnalysis(context, librarySource, sourceFile, lineInfoMap, errors);
  }

  protected ErrorSeverity getMaxErrorSeverity(List<AnalysisError> errors) {
    ErrorSeverity status = ErrorSeverity.NONE;

    for (AnalysisError error : errors) {
      ErrorSeverity severity = error.getErrorCode().getErrorSeverity();

      status = status.max(severity);
    }

    return status;
  }

  protected ErrorSeverity performAnalysis(AnalysisContext context, Source librarySource,
      File sourceFile, Map<Source, LineInfo> lineInfoMap, List<AnalysisError> errors)
      throws AnalysisException {
    // don't try to analyze parts
    if (context.computeKindOf(librarySource) == SourceKind.PART) {
      System.err.println("Only libraries can be analyzed.");
      System.err.println(sourceFile + " is a part and can not be analyzed.");
      return ErrorSeverity.NONE;
    }

    // analyze Source
    LibraryElement library = context.computeLibraryElement(librarySource);
    context.resolveCompilationUnit(librarySource, library);

    // prepare errors
    Set<Source> sources = getAllSources(library);
    getAllErrors(context, sources, errors, lineInfoMap);
    filterOutTodos(errors);
    if (options.getDisableHints()) {
      filterOutHints(errors);
    }
    return getMaxErrorSeverity(errors);
  }

  Set<Source> getAllSources(LibraryElement library) {
    Set<CompilationUnitElement> units = new HashSet<CompilationUnitElement>();
    Set<LibraryElement> libraries = new HashSet<LibraryElement>();
    Set<Source> sources = new HashSet<Source>();

    addLibrary(library, libraries, units, sources);

    return sources;
  }

  private void addCompilationUnit(CompilationUnitElement unit, Set<LibraryElement> libraries,
      Set<CompilationUnitElement> units, Set<Source> sources) {
    if (unit == null || units.contains(unit)) {
      return;
    }

    units.add(unit);

    sources.add(unit.getSource());
  }

  private void addLibrary(LibraryElement library, Set<LibraryElement> libraries,
      Set<CompilationUnitElement> units, Set<Source> sources) {
    if (library == null || libraries.contains(library)) {
      return;
    }

    UriKind uriKind = library.getSource().getUriKind();

    // Optionally skip package: libraries.
    if (!options.getShowPackageWarnings() && uriKind == UriKind.PACKAGE_URI) {
      return;
    }

    // Optionally skip SDK libraries.
    if (!options.getShowSdkWarnings() && uriKind == UriKind.DART_URI) {
      return;
    }

    libraries.add(library);

    // add compilation units
    addCompilationUnit(library.getDefiningCompilationUnit(), libraries, units, sources);

    for (CompilationUnitElement child : library.getParts()) {
      addCompilationUnit(child, libraries, units, sources);
    }

    // add referenced libraries
    for (LibraryElement child : library.getImportedLibraries()) {
      addLibrary(child, libraries, units, sources);
    }

    for (LibraryElement child : library.getExportedLibraries()) {
      addLibrary(child, libraries, units, sources);
    }
  }

  /**
   * Create the source factory to be used in the analysis context.
   * 
   * @param sourceFile the file to be analyzed
   * @return the source factory that was created
   */
  private SourceFactory createSourceFactory(File sourceFile) {
    File packageDirectory = getPackageDirectory(sourceFile);
    if (options.getUsePackageMap()) {
      return new SourceFactory(
          new DartUriResolver(sdk),
          new FileUriResolver(),
          new ExplicitPackageUriResolver(sdk, getPubDir(sourceFile)));
    } else if (packageDirectory != null) {
      return new SourceFactory(
          new DartUriResolver(sdk),
          new FileUriResolver(),
          new PackageUriResolver(packageDirectory.getAbsoluteFile()));
    } else {
      return new SourceFactory(new DartUriResolver(sdk), new FileUriResolver());
    }
  }

  /**
   * Remove any hints (ErrorType.HINT) from the passed list.
   */
  private void filterOutHints(List<AnalysisError> errors) {
    for (int i = errors.size() - 1; i >= 0; i--) {
      AnalysisError error = errors.get(i);
      if (error.getErrorCode().getType() == ErrorType.HINT) {
        errors.remove(i);
      }
    }
  }

  /**
   * Remove any to-do's (ErrorType.TODO) from the passed list.
   */
  private void filterOutTodos(List<AnalysisError> errors) {
    for (int i = errors.size() - 1; i >= 0; i--) {
      AnalysisError error = errors.get(i);
      if (error.getErrorCode().getType() == ErrorType.TODO) {
        errors.remove(i);
      }
    }
  }

  private void getAllErrors(AnalysisContext context, Set<Source> sources,
      List<AnalysisError> errors, Map<Source, LineInfo> lineInfoMap) throws AnalysisException {
    for (Source source : sources) {
      AnalysisError[] sourceErrors = context.computeErrors(source);
      if (sourceErrors.length > 0) {
        errors.addAll(Arrays.asList(sourceErrors));
        LineInfo lineInfo = context.getLineInfo(source);
        if (lineInfo == null) {
          lineInfo = new LineInfo(new int[] {0});
        }
        lineInfoMap.put(source, lineInfo);
      }
    }
  }

  /**
   * Return the package directory to be used to resolve {@code package:} URI's.
   * 
   * @param sourceFile the file to be analyzed
   * @return the package directory to be used to resolve {@code package:} URI's
   */
  private File getPackageDirectory(File sourceFile) {
    if (options.getPackageRootPath() != null) {
      return options.getPackageRootPath();
    } else {
      return getPackageDirectoryFor(sourceFile);
    }
  }

  private File getPackageDirectoryFor(File sourceFile) {
    // we are going to ask parent file, so get absolute path
    sourceFile = sourceFile.getAbsoluteFile();

    // look in the containing directories
    File dir = sourceFile.getParentFile();
    while (dir != null) {
      File packagesDir = new File(dir, "packages");
      if (packagesDir.exists()) {
        return packagesDir;
      }
      dir = dir.getParentFile();
    }

    return null;
  }

  /**
   * Return a directory containing a pubspec.yaml file. The search location starts at the parent of
   * the given source file and continues up the tree. If no pub directory is found, return the cwd.
   * 
   * @param sourceFile the starting location
   * @return a directory containing a pubspec.yaml file, or the cwd if no such directory is found
   */
  private File getPubDir(File sourceFile) {
    // we are going to ask parent file, so get absolute path
    sourceFile = sourceFile.getAbsoluteFile();

    // look in the containing directories
    File dir = sourceFile.getParentFile();

    while (dir != null) {
      File pubspecFile = new File(dir, "pubspec.yaml");

      if (pubspecFile.exists()) {
        return dir;
      }

      dir = dir.getParentFile();
    }

    // Else, return the cwd.
    return new File(System.getProperty("user.dir"));
  }

  /**
   * Returns the URI for the given input file. This is usually a file: URI, but if the given file is
   * located in the "lib" directory of the {@link #sdk}, then this method returns a dart URI.
   */
  private URI getUri(File file) {
    // may be file in SDK
    if (sdk instanceof DirectoryBasedDartSdk) {
      DirectoryBasedDartSdk directoryBasedSdk = sdk;
      File libraryDirectory = directoryBasedSdk.getLibraryDirectory();
      String sdkLibPath = libraryDirectory.getPath() + File.separator;
      String filePath = file.getPath();
      if (filePath.startsWith(sdkLibPath)) {
        String internalPath = new File(libraryDirectory, "_internal").getPath() + File.separator;
        if (!filePath.startsWith(internalPath)) {
          Source source = sdk.fromFileUri(file.toURI());
          if (source != null) {
            return source.getUri();
          }
        }
      }
    }
    // some generic file
    return file.toURI();
  }
}

File: dart/pkg/analyzer/lib/src/analyzer_impl.dart
// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library analyzer_impl;

import 'dart:async';
import 'dart:collection';
import 'dart:io';

import 'package:analyzer/file_system/file_system.dart' show Folder;
import 'package:analyzer/file_system/physical_file_system.dart';
import 'package:analyzer/source/package_map_provider.dart';
import 'package:analyzer/source/package_map_resolver.dart';
import 'package:analyzer/source/pub_package_map_provider.dart';
import 'package:analyzer/src/error_formatter.dart';
import 'package:analyzer/src/generated/java_core.dart' show JavaSystem;
import 'package:analyzer/src/generated/java_engine.dart';

import '../options.dart';
import 'generated/constant.dart';
import 'generated/element.dart';
import 'generated/engine.dart';
import 'generated/error.dart';
import 'generated/java_io.dart';
import 'generated/sdk_io.dart';
import 'generated/source_io.dart';
import 'package:analyzer/src/generated/utilities_general.dart';

DirectoryBasedDartSdk sdk;

/**
 * The maximum number of sources for which AST structures should be kept in the cache.
 */
const int _MAX_CACHE_SIZE = 512;

/// Analyzes single library [File].
class AnalyzerImpl {
  final String sourcePath;

  final CommandLineOptions options;
  final int startTime;
  ContentCache contentCache = new ContentCache();

  SourceFactory sourceFactory;
  AnalysisContext context;
  Source librarySource;
  /// All [Source]s references by the analyzed library.
  final Set<Source> sources = new Set<Source>();

  /// All [AnalysisErrorInfo]s in the analyzed library.
  final List<AnalysisErrorInfo> errorInfos = new List<AnalysisErrorInfo>();

  /// [HashMap] between sources and analysis error infos.
  final HashMap<Source, AnalysisErrorInfo> sourceErrorsMap =
      new HashMap<Source, AnalysisErrorInfo>();

  AnalyzerImpl(String sourcePath, this.options, this.startTime)
      : sourcePath = _normalizeSourcePath(sourcePath) {
    if (sdk == null) {
      sdk = new DirectoryBasedDartSdk(new JavaFile(options.dartSdkPath));
    }
  }

  /// Returns the maximal [ErrorSeverity] of the recorded errors.
  ErrorSeverity get maxErrorSeverity {
    var status = ErrorSeverity.NONE;
    for (AnalysisErrorInfo errorInfo in errorInfos) {
      for (AnalysisError error in errorInfo.errors) {
        if (!_isDesiredError(error)) {
          continue;
        }
        var severity = computeSeverity(error, options.enableTypeChecks);
        status = status.max(severity);
      }
    }
    return status;
  }

  void addCompilationUnitSource(CompilationUnitElement unit,
      Set<LibraryElement> libraries, Set<CompilationUnitElement> units) {
    if (unit == null || units.contains(unit)) {
      return;
    }
    units.add(unit);
    sources.add(unit.source);
  }

  void addLibrarySources(LibraryElement library, Set<LibraryElement> libraries,
      Set<CompilationUnitElement> units) {
    if (library == null || !libraries.add(library)) {
      return;
    }
    // may be skip library
    {
      UriKind uriKind = library.source.uriKind;
      // Optionally skip package: libraries.
      if (!options.showPackageWarnings && uriKind == UriKind.PACKAGE_URI) {
        return;
      }
      // Optionally skip SDK libraries.
      if (!options.showSdkWarnings && uriKind == UriKind.DART_URI) {
        return;
      }
    }
    // add compilation units
    addCompilationUnitSource(library.definingCompilationUnit, libraries, units);
    for (CompilationUnitElement child in library.parts) {
      addCompilationUnitSource(child, libraries, units);
    }
    // add referenced libraries
    for (LibraryElement child in library.importedLibraries) {
      addLibrarySources(child, libraries, units);
    }
    for (LibraryElement child in library.exportedLibraries) {
      addLibrarySources(child, libraries, units);
    }
  }

  /**
   * Treats the [sourcePath] as the top level library and analyzes it using a
   * asynchronous algorithm over the analysis engine.
   */
  void analyzeAsync() {
    setupForAnalysis();
    _analyzeAsync();
  }

  /**
   * Treats the [sourcePath] as the top level library and analyzes it using a
   * synchronous algorithm over the analysis engine. If [printMode] is `0`,
   * then no error or performance information is printed. If [printMode] is `1`,
   * then both will be printed. If [printMode] is `2`, then only performance
   * information will be printed, and it will be marked as being for a cold VM.
   */
  ErrorSeverity analyzeSync({int printMode: 1}) {
    setupForAnalysis();
    return _analyzeSync(printMode);
  }

  void prepareAnalysisContext(JavaFile sourceFile, Source source) {
    List<UriResolver> resolvers = [
        new CustomUriResolver(options.customUrlMappings),
        new DartUriResolver(sdk),
        new FileUriResolver()];
    // may be add package resolver
    {
      JavaFile packageDirectory;
      if (options.packageRootPath != null) {
        packageDirectory = new JavaFile(options.packageRootPath);
        resolvers.add(new PackageUriResolver([packageDirectory]));
      } else {
        PubPackageMapProvider pubPackageMapProvider =
            new PubPackageMapProvider(PhysicalResourceProvider.INSTANCE, sdk);
        PackageMapInfo packageMapInfo = pubPackageMapProvider.computePackageMap(
            PhysicalResourceProvider.INSTANCE.getResource('.'));
        Map<String, List<Folder>> packageMap = packageMapInfo.packageMap;
        if (packageMap != null) {
          resolvers.add(
              new PackageMapUriResolver(
                  PhysicalResourceProvider.INSTANCE,
                  packageMap));
        }
      }
    }
    sourceFactory = new SourceFactory(resolvers);
    context = AnalysisEngine.instance.createAnalysisContext();
    context.sourceFactory = sourceFactory;
    Map<String, String> definedVariables = options.definedVariables;
    if (!definedVariables.isEmpty) {
      DeclaredVariables declaredVariables = context.declaredVariables;
      definedVariables.forEach((String variableName, String value) {
        declaredVariables.define(variableName, value);
      });
    }
    // Uncomment the following to have errors reported on stdout and stderr
    AnalysisEngine.instance.logger = new StdLogger(options.log);

    // set options for context
    AnalysisOptionsImpl contextOptions = new AnalysisOptionsImpl();
    contextOptions.cacheSize = _MAX_CACHE_SIZE;
    contextOptions.hint = !options.disableHints;
    context.analysisOptions = contextOptions;

    // Create and add a ChangeSet
    ChangeSet changeSet = new ChangeSet();
    changeSet.addedSource(source);
    context.applyChanges(changeSet);
  }

  /// Fills [errorInfos] using [sources].
  void prepareErrors() {
    for (Source source in sources) {
      context.computeErrors(source);
      var sourceErrors = context.getErrors(source);
      errorInfos.add(sourceErrors);
    }
  }

  /// Fills [sources].
  void prepareSources(LibraryElement library) {
    var units = new Set<CompilationUnitElement>();
    var libraries = new Set<LibraryElement>();
    addLibrarySources(library, libraries, units);
  }

  /**
   * Setup local fields such as the analysis context for analysis.
   */
  void setupForAnalysis() {
    sources.clear();
    errorInfos.clear();
    if (sourcePath == null) {
      throw new ArgumentError("sourcePath cannot be null");
    }
    JavaFile sourceFile = new JavaFile(sourcePath);
    Uri uri = getUri(sourceFile);
    librarySource = new FileBasedSource.con2(uri, sourceFile);

    // prepare context
    prepareAnalysisContext(sourceFile, librarySource);
  }

  /// The async version of the analysis
  void _analyzeAsync() {
    new Future(context.performAnalysisTask).then((AnalysisResult result) {
      List<ChangeNotice> notices = result.changeNotices;
      if (result.hasMoreWork) {
        // There is more work, record the set of sources, and then call self
        // again to perform next task
        for (ChangeNotice notice in notices) {
          sources.add(notice.source);
          sourceErrorsMap[notice.source] = notice;
        }
        return _analyzeAsync();
      }
      //
      // There are not any more tasks, set error code and print performance
      // numbers.
      //
      // prepare errors
      sourceErrorsMap.forEach((k, v) {
        errorInfos.add(sourceErrorsMap[k]);
      });

      // print errors and performance numbers
      _printErrorsAndPerf();

      // compute max severity and set exitCode
      ErrorSeverity status = maxErrorSeverity;
      if (status == ErrorSeverity.WARNING && options.warningsAreFatal) {
        status = ErrorSeverity.ERROR;
      }
      exitCode = status.ordinal;
    }).catchError((ex, st) {
      AnalysisEngine.instance.logger.logError("$ex\n$st");
    });
  }

  /// The sync version of analysis.
  ErrorSeverity _analyzeSync(int printMode) {
    // don't try to analyze parts
    if (context.computeKindOf(librarySource) == SourceKind.PART) {
      print("Only libraries can be analyzed.");
      print("$sourcePath is a part and can not be analyzed.");
      return ErrorSeverity.ERROR;
    }
    // resolve library
    var libraryElement = context.computeLibraryElement(librarySource);
    // prepare source and errors
    prepareSources(libraryElement);
    prepareErrors();

    // print errors and performance numbers
    if (printMode == 1) {
      _printErrorsAndPerf();
    } else if (printMode == 2) {
      _printColdPerf();
    }

    // compute max severity and set exitCode
    ErrorSeverity status = maxErrorSeverity;
    if (status == ErrorSeverity.WARNING && options.warningsAreFatal) {
      status = ErrorSeverity.ERROR;
    }
    return status;
  }

  bool _isDesiredError(AnalysisError error) {
    if (error.errorCode.type == ErrorType.TODO) {
      return false;
    }
    if (computeSeverity(error, options.enableTypeChecks) ==
        ErrorSeverity.INFO &&
        options.disableHints) {
      return false;
    }
    return true;
  }

  _printColdPerf() {
    // print cold VM performance numbers
    int totalTime = JavaSystem.currentTimeMillis() - startTime;
    int otherTime = totalTime;
    for (PerformanceTag tag in PerformanceTag.all) {
      if (tag != PerformanceTag.UNKNOWN) {
        int tagTime = tag.elapsedMs;
        stdout.writeln('${tag.label}-cold:$tagTime');
        otherTime -= tagTime;
      }
    }
    stdout.writeln('other-cold:$otherTime');
    stdout.writeln("total-cold:$totalTime");
  }

  _printErrorsAndPerf() {
    // The following is a hack. We currently print out to stderr to ensure that
    // when in batch mode we print to stderr, this is because the prints from
    // batch are made to stderr. The reason that options.shouldBatch isn't used
    // is because when the argument flags are constructed in BatchRunner and
    // passed in from batch mode which removes the batch flag to prevent the
    // "cannot have the batch flag and source file" error message.
    IOSink sink = options.machineFormat ? stderr : stdout;

    // print errors
    ErrorFormatter formatter =
        new ErrorFormatter(sink, options, _isDesiredError);
    formatter.formatErrors(errorInfos);

    // print performance numbers
    if (options.perf || options.warmPerf) {
      int totalTime = JavaSystem.currentTimeMillis() - startTime;
      int otherTime = totalTime;
      for (PerformanceTag tag in PerformanceTag.all) {
        if (tag != PerformanceTag.UNKNOWN) {
          int tagTime = tag.elapsedMs;
          stdout.writeln('${tag.label}:$tagTime');
          otherTime -= tagTime;
        }
      }
      stdout.writeln('other:$otherTime');
      stdout.writeln("total:$totalTime");
    }
  }

  /**
   * Compute the severity of the error; however, if
   * [enableTypeChecks] is false, then de-escalate checked-mode compile time
   * errors to a severity of [ErrorSeverity.INFO].
   */
  static ErrorSeverity computeSeverity(AnalysisError error,
      bool enableTypeChecks) {
    if (!enableTypeChecks &&
        error.errorCode.type == ErrorType.CHECKED_MODE_COMPILE_TIME_ERROR) {
      return ErrorSeverity.INFO;
    }
    return error.errorCode.errorSeverity;
  }

  static JavaFile getPackageDirectoryFor(JavaFile sourceFile) {
    // we are going to ask parent file, so get absolute path
    sourceFile = sourceFile.getAbsoluteFile();
    // look in the containing directories
    JavaFile dir = sourceFile.getParentFile();
    while (dir != null) {
      JavaFile packagesDir = new JavaFile.relative(dir, "packages");
      if (packagesDir.exists()) {
        return packagesDir;
      }
      dir = dir.getParentFile();
    }
    // not found
    return null;
  }

  /**
   * Returns the [Uri] for the given input file.
   *
   * Usually it is a `file:` [Uri], but if [file] is located in the `lib`
   * directory of the [sdk], then returns a `dart:` [Uri].
   */
  static Uri getUri(JavaFile file) {
    // may be file in SDK
    {
      Source source = sdk.fromFileUri(file.toURI());
      if (source != null) {
        return source.uri;
      }
    }
    // some generic file
    return file.toURI();
  }

  /**
   * Convert [sourcePath] into an absolute path.
   */
  static String _normalizeSourcePath(String sourcePath) {
    return new File(sourcePath).absolute.path;
  }
}

/**
 * This [Logger] prints out information comments to [stdout] and error messages
 * to [stderr].
 */
class StdLogger extends Logger {
  final bool log;

  StdLogger(this.log);

  @override
  void logError(String message, [CaughtException exception]) {
    stderr.writeln(message);
    if (exception != null) {
      stderr.writeln(exception);
    }
  }

  @override
  void logError2(String message, Object exception) {
    stderr.writeln(message);
  }

  @override
  void logInformation(String message, [CaughtException exception]) {
    if (log) {
      stdout.writeln(message);
      if (exception != null) {
        stderr.writeln(exception);
      }
    }
  }

  @override
  void logInformation2(String message, Object exception) {
    if (log) {
      stdout.writeln(message);
    }
  }
}
